log/test/loggers.go

```diff
@@ -2,12 +2,99 @@ package test
 
 import (
 	"fmt"
+	"io/ioutil"
+	"os"
+	"strings"
 	"sync"
 	"testing"
+
+	"github.com/tendermint/tendermint/libs/log"
 )
 
 // TODO(tzdybal): move to some common place
 
+// FileLogger is a simple, thread-safe logger that logs to a tempFile and is
+// intended for use in testing.
+type FileLogger struct {
+	kv      []interface{}
+	logFile *os.File
+	T       *testing.T
+	mtx     *sync.Mutex
+}
+
+// NewLogger create a Logger using the name of the test as the filename.
+func NewFileLogger(t *testing.T) *FileLogger {
+	// Use the test name but strip out any slashes since they are not
+	// allowed in filenames.
+	return NewFileLoggerCustom(t, strings.ReplaceAll(t.Name(), "/", "_"))
+}
+
+// NewLoggerCustom create a Logger using the given filename.
+func NewFileLoggerCustom(t *testing.T, fileName string) *FileLogger {
+	logFile, err := ioutil.TempFile("", fileName)
```

Comment:
I get the idea, but I don't like that files are polluting my drive locally. Actually, maybe using something more deterministic would be better - like:
```go
// NewFileLogger create a Logger that writes to temporary file.
func NewFileLogger(t *testing.T) log.Logger {
	logFile, err := os.CreateTemp(t.TempDir(), "log")
	if err != nil {
		panic(err)
	}
	t.Log("Test logs can be found:", logFile.Name())
	t.Cleanup(func() {
		_ = logFile.Close()
	})
	return log.NewTMLogger(logFile)
}
```

This way you can still stop the test in debugger (or add `os.Exit(1)`) and read the logs, but they wont accumulate.

Reply:
I don't think we want to do this as the `t.TempDir` is removed on testing clean up. So we wouldn't be able to access the logs. 

The current implementation puts the log files in the system temp directory which is cleaned up whenever you shutdown your computer so you aren't polluting your drive. 

Unless you know how to access the files in the testing TempDir after the test is complete, I don't think is achieves the result we are looking for.

Reply:
As I said before, this requires putting breakpoint at the end of test in debugger, or calling `os.Exit` to ensure that test cleanup won't run. 

So my perspective is - as a developer, when I struggle with a test and I need more debugging information, there's a way to get it. 

The plus of your solution is, after we spot a flaky test, we can do the postmortem analysis with all the data. This is definitely great for CI.

Reply:
yes